app/src/main/java/org/greenstand/android/TreeTracker/login/LoginView.kt

```diff
@@ -0,0 +1,221 @@
+
+                        viewModel.loginButtonClicked(credentials)
+
+                        textInputService?.hideSoftwareKeyboard()
+                    }
+                )
+
+                LoginButton(
+                    emailTextState = emailTextInputState,
+                    phoneTextState = phoneTextInputState,
+                    onClick = {
+                        val credentials = if (emailTextInputState.value.isNotBlank()) {
+                            emailTextInputState.value
+                        } else {
+                            phoneTextInputState.value
+                        }
+
+                        viewModel.loginButtonClicked(credentials)
+                    }
+                )
+            }
+        }
+    }
+}
+
+@Composable
+private fun PhoneInputText(
+    textInputState: MutableState<String>,
+    onActionDone: KeyboardActionScope.() -> Unit
+) {
+    TextField(
+        modifier = Modifier
+            .padding(top = 28.dp)
+            .fillMaxWidth()
+            .wrapContentHeight(),
+        value = textInputState.value,
+        onValueChange = { newInput ->
+            textInputState.value = newInput
+        },
+        label = {
+            Text(text = stringResource(id = R.string.phone_text_hint))
+        },
+        leadingIcon = {
+            Icon(
+                modifier = Modifier.padding(16.dp),
+                imageVector = ImageVector.vectorResource(id = R.drawable.phone),
+                contentDescription = stringResource(id = R.string.content_description_phone)
+            )
+        },
+        keyboardOptions = KeyboardOptions(
+            keyboardType = KeyboardType.Phone,
+            imeAction = ImeAction.Done
+        ),
+        keyboardActions = KeyboardActions(
+            onDone = onActionDone
+        )
+
+    )
+}
+
+@Composable
+private fun EmailInputText(
+    textInputState: MutableState<String>,
+    onActionGo: KeyboardActionScope.() -> Unit
+) {
+    TextField(
+        modifier = Modifier
+            .padding(top = 8.dp)
+            .fillMaxWidth()
+            .wrapContentHeight(),
+        value = textInputState.value,
+        onValueChange = { newInput ->
+            textInputState.value = newInput
+        },
+        label = {
+            Text(text = stringResource(id = R.string.email_text_hint))
+        },
+        leadingIcon = {
+            Icon(
+                modifier = Modifier.padding(16.dp),
+                imageVector = ImageVector.vectorResource(id = R.drawable.email),
+                contentDescription = stringResource(id = R.string.content_description_email)
+            )
+        },
+        keyboardOptions = KeyboardOptions(
+            keyboardType = KeyboardType.Email,
+            imeAction = ImeAction.Go
+        ),
+        keyboardActions = KeyboardActions(
+            onGo = onActionGo
+        )
+
+    )
+}
+
+@Composable
+private fun LoginButton(
+    emailTextState: MutableState<String>,
```

Comment:
Is there a reason we don't just pass strings to this method?

Reply:
Yeah. So Compose updates it's views whenever its `State` changes. By utilizing the `State<T>` interface (`MutableState<T>` implements this interface). So by attaching the email and phone states to the `isEnabled` argument it will ensure that the button enables and disables on it's own without us having to do anything :) it's all handled for us!

Reply:
I realize that my explanation my have been superfluous! My apologies, I was more so thinking outloud in what I have understood how the UI updates and passing just the text I think wouldn't update the button enabled state in real-time due to the whole state thing